core/tee/tee_enc_fs.c

```diff
@@ -0,0 +1,761 @@
+			encrypted_data, &encrypted_data_len);
+	if (ret != TEE_SUCCESS)
+		goto exit_free_encrypted_data;
+
+	res = lseek(fdp->nw_fd, 0, TEE_FS_SEEK_SET);
+	if (res < 0)
+		goto exit_free_encrypted_data;
+
+	res = ftruncate(fdp->nw_fd, encrypted_data_len);
+	if (res < 0)
+		goto exit_free_encrypted_data;
+
+	/* write encrypted file content to normal world file system */
+	res = write_file(fdp->nw_fd, encrypted_data, encrypted_data_len);
+	if (res != (int)encrypted_data_len) {
+		res = -1;
+		goto exit_free_encrypted_data;
+	}
+	DMSG("%d bytes written", res);
+
+	/* fill in parameters */
+	head.op = TEE_FS_CLOSE;
+	head.fd = fdp->nw_fd;
+
+	res = tee_fs_send_cmd(&head, NULL, 0, TEE_FS_MODE_NONE);
+	if (res != 0)
+		goto exit_free_encrypted_data;
+
+	res = head.res;
+
+exit_free_encrypted_data:
+	free(encrypted_data);
+exit_free_fd:
+	free(fdp->data);
+	free(fdp);
+
+	return res;
+}
+
+static int tee_fs_read(int fd, void *buf, size_t len)
+{
+	int res = -1;
+	size_t remain;
+	struct tee_fs_fd *fdp = handle_lookup(&fs_handle_db, fd);
+
+	DMSG("read, fd=%d, buf=%p, len=%d", fd, buf, len);
+
+	if (len == 0) {
+		res = 0;
+		goto exit;
+	}
+
+	if (fdp == NULL || buf == NULL)
+		goto exit;
+
+	if (fdp->flags & TEE_FS_O_WRONLY)
+		goto exit;
+
+	remain = fdp->len - fdp->pos;
+	if (len > remain)
+		len = remain;
+
+	memcpy(buf, fdp->data + fdp->pos, len);
+	res = len;
+
+exit:
+	return res;
+}
+
+static int tee_fs_write(int fd, const void *buf, size_t len)
+{
+	int res = -1;
+	size_t remain;
+	struct tee_fs_fd *fdp = handle_lookup(&fs_handle_db, fd);
+
+	DMSG("write, fd=%d, buf=%p, len=%d", fd, buf, len);
+
+	if (len == 0) {
+		res = 0;
+		goto exit;
+	}
+
+	if (buf == NULL)
+		goto exit;
+
+	if (fdp == NULL)
+		goto exit;
+
+	/* restrict this to avoid file hole */
+	if (fdp->pos > fdp->len)
+		goto exit;
+
+	if (fdp->flags & TEE_FS_O_RDONLY)
+		goto exit;
+
+	/* check if we need to update file size */
+	remain = fdp->len - fdp->pos;
+	if (len > remain) {
+		size_t new_file_len = fdp->len + (len - remain);
+
+		res = update_file_size(fdp, new_file_len);
+		if (res != 0)
+			goto exit;
+	}
+
+	memcpy(fdp->data + fdp->pos, buf, len);
+	fdp->pos += len;
+	res = len;
+
+exit:
+	return res;
+}
+
+static tee_fs_off_t tee_fs_lseek(int fd, tee_fs_off_t offset, int whence)
+{
+	tee_fs_off_t res = -1;
+	tee_fs_off_t new_pos;
+	struct tee_fs_fd *fdp = handle_lookup(&fs_handle_db, fd);
+
+	if (!fdp)
+		goto exit;
+
+	switch (whence) {
+
+	case TEE_FS_SEEK_SET:
+		new_pos = offset;
+		break;
+
+	case TEE_FS_SEEK_CUR:
+		new_pos = fdp->pos + offset;
+		break;
+
+	case TEE_FS_SEEK_END:
+		new_pos = fdp->len + offset;
+		break;
+
+	default:
+		goto exit;
+	}
+
+	/*
+	 * file hole is not supported in this implementation,
+	 * restrict the file postion within file length
+	 * for simplicity
+	 */
+	if (new_pos > (tee_fs_off_t)fdp->len)
+		goto exit;
+
+	res = fdp->pos = new_pos;
+exit:
+	return res;
+}
+
+static int tee_fs_rename(const char *old, const char *new)
```

Comment:
Oh, I have to admit I didn't know that this was way we did this in tee-supplicant. Seems pretty strange way to handle it in the same buffer. But since that is how it's done I don't have any other comments.